### ISSUE TITLE
Ergonomic ref-counting

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/goals#107]
| Metadata         | |
| --------         | --- |
| Point of contact | @nikomatsakis |
| Team(s)          | [compiler](http://github.com/rust-lang/compiler-team), [lang](http://github.com/rust-lang/lang-team), [lang-docs](https://www.rust-lang.org/governance/teams), [libs-api](https://www.rust-lang.org/governance/teams) |
| Goal document    | [2026/ergonomic-rc](https://rust-lang.github.io/goals/2026/ergonomic-rc.html) |

## Summary

Implement and prototype two foundational improvements for ergonomic ref-counting: (1) a `Share` trait that semantically identifies types where cloning creates an alias to the same underlying value, and (2) move expressions (`move($expr)`) that allow precise control over what closures capture and when. These changes lay groundwork for future ergonomic improvements while delivering immediate value, with prototypes targeted for summer 2026.

## Tasks and status

[Team]: https://img.shields.io/badge/Team%20ask-red

---
Note: we have updated the body to match the 2026 goal. Your original text is preserved below. <details>

| Metadata         | |
| --------         | --- |
| Point of contact | @nikomatsakis |
| Team(s)          | [compiler](http://github.com/rust-lang/compiler-team), [lang](http://github.com/rust-lang/lang-team), [lang-docs](https://www.rust-lang.org/governance/teams), [libs-api](https://www.rust-lang.org/governance/teams) |
| Goal document    | [2026/ergonomic-rc](https://rust-lang.github.io/rust-project-goals/2026/ergonomic-rc.html) |

## Summary

Implement and prototype two foundational improvements for ergonomic ref-counting: (1) a `Share` trait that semantically identifies types where cloning creates an alias to the same underlying value, and (2) move expressions (`move($expr)`) that allow precise control over what closures capture and when. These changes lay groundwork for future ergonomic improvements while delivering immediate value, with prototypes targeted for summer 2026.

**Needs contributor:** This is a medium complexity problem. @nikomatsakis would like to mentor this goal with either an experienced contributor or a cohort of folks who know Rust but not the compiler. It is estimated that the goal will take about 6 months to implement for a cohort with minimal compiler hacking experience.

## Tasks and status

[Team]: https://img.shields.io/badge/Team%20ask-red

---
Note: we have updated the body to match the 2026 goal. Your original text is preserved below. <details>

| Metadata         | |
| --------         | --- |
| Point of contact | @nikomatsakis |
| Team(s)          | [compiler](http://github.com/rust-lang/compiler-team), [lang](http://github.com/rust-lang/lang-team) |
| Goal document    | [2025h2/ergonomic-rc](https://rust-lang.github.io/rust-project-goals/2025h2/ergonomic-rc.html) |

## Summary

We propose to write an alternative RFC for ergonomic ref-counting that makes lightweight cloning automatic and hold design meetings so the lang team can compare both approaches. This work builds on RFC #3680, which proposed a new keyword, `use`, that could be used with closures (`use || ...`) and expressions like `x.use` to help address a longstanding problem: working with ref-counted data structures like `Arc<T>` is verbose and confusing.

The 2025H1 work delivered `.use` syntax that works technically, but community feedback on the RFC pointed out a tension: if we're trying to improve ergonomics, why add more required syntax? We'll write an RFC exploring the automatic approach, complete the implementation work needed to support it, and run design meetings to help the lang team decide. This will tell us about the impact on user experience and performance, how well we can catch problematic cases, and which approach works better for Rust developers working with ref-counted data.

## Tasks and status

* [ ] Author alternative RFC (@nikomatsakis)
* [ ] Complete seamless implementation (@spastorino)
* [ ] Standard reviews ([compiler](http://github.com/rust-lang/compiler-team) ![Team][])
* [ ] Design meeting ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [ ] RFC decision ([lang](http://github.com/rust-lang/lang-team) ![Team][])

[Team]: https://img.shields.io/badge/Team%20ask-red

---
Note: we have updated the body to match the 2025h2 goal. Your original text is preserved below. <details>

| Metadata         | |
| --------         | --- |
| Point of contact | @spastorino |
| Team(s)          | [lang](http://github.com/rust-lang/lang-team) |
| Goal document    | [2025h1/ergonomic-rc](https://rust-lang.github.io/rust-project-goals/2025h1/ergonomic-rc.html) |

## Summary

* Deliver a nightly implementation of the experimental `use` syntax for ergonomic ref-counting.
* RFC decision on the above

## Tasks and status

* [x] Implementation (@spastorino)
* [x] Reviews (@nikomatsakis)
* [x] Lang-team champion ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [x] Author RFC (@joshtriplett)
* [x] Introduce `x.use` and `use ||` syntax and `UseCloned` trait (rust-lang/rust#134797).
* [x] Modify codegen so that we guarantee that `x.use` will do a copy if `X: Copy` is true after monomorphization. Right now the desugaring to `clone` occurs before monomorphization and hence it will call the `clone` method even for those instances where `X` is a `Copy` type.
* [ ] Convert `x.use` to a move rather than a clone if this is a last-use.
* [ ] Make `x` equivalent to `x.use` but with an (allow-by-default) lint to signal that something special is happened.

[Team]: https://img.shields.io/badge/Team%20ask-red

</details>
</details>
</details>

**Comment by nikomatsakis** [comment]
This issue has not seen many updates, but there has been definite progress:

* @joshtriplett opened https://github.com/rust-lang/rfcs/pull/3680. The @rust-lang/lang team has not yet truly discussed or reached a decision on that RFC.
* @spastorino began implementation work on a prototype.

**Comment by nikomatsakis** [comment]
Year-end summary:

In 2024H2 @spastorino began work on an experimental implementation (not yet landed) and @joshtriplett authored a corresponding RFC (rust-lang/rfcs#3680), which has received substantial feedback.

We opened a proposed 2025H1 project goal to continue by (a) landing the experimental branch and (b) addressing feedback on the RFC, reading it with the lang-team, and reaching a decision.

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h1

**Comment by spastorino** [comment]
There's a PR up https://github.com/rust-lang/rust/pull/134797 which implements the proposed RFC without the optimizations.
The PR is not yet merged and we need to continue now working on addressing comments to the PR until is merged and then start implementing optimizations.

**Comment by nikomatsakis** [comment]
Update: rust-lang/rust#134797 has landed.

Semantics as implemented in the PR:

* [x] Introduced a trait `UseCloned` implemented for `Rc` and `Arc` types.
* [x] `x.use` checks whether `x`'s type `X` implements the `UseCloned` trait; if so, then `x.use` is equivalent to `x.clone()`, otherwise it is a copy/move of `x`;
* [x] `use || ...x...` closures act like `move` closures but respect the `UseCloned` trait, so they will either `clone`, copy, or move `x` as appropriate.

Next steps:

* [ ] Modify codegen so that we guarantee that `x.use` will do a copy if `X: Copy` is true after monomorphization. Right now the desugaring to `clone` occurs before monomorphization and hence it will call the `clone` method even for those instances where `X` is a `Copy` type.
* [ ] Convert `x.use` to a move rather than a clone if this is a last-use.
* [ ] Make `x` equivalent to `x.use` but with an (allow-by-default) lint to signal that something special is happened.

Notable decisions made and discussions: 

* Opted to name the trait that controls whether `x.use` does a clone or a move `UseCloned` rather than `Use`. This is because the trait does not control whether or not you can use something but rather controls what happens when you do.
* [Question was raised on Zulip](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/.60ergonomic_clones.60.20does.20not.20deref/near/505889669) as to whether `x.use` should auto-deref. After thinking it over, reached the conclusion that [it should not](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/.60ergonomic_clones.60.20does.20not.20deref/near/506157506), because `x` and `x.use` should eventually behave the same modulo lints, but that (as ever) a `&T -> T` coercion would be useful for ergonomic reasons.

**Comment by nikomatsakis** [comment]
TL;DR: There is now an early prototype available that allows you to write `x.use`; if the type of `X` implements `UseCloned`, then this is equivalent to `x.clone()`, else it is equivalent to a move. This is not the desired end semantics in a few ways, just a step along the road. Nothing to see here (yet).

**Comment by nikomatsakis** [comment]
In reviewing https://github.com/rust-lang/rust/pull/138628, we realized that the tests were not behaving as expected because they were running in Rust 2015 which had distinct capture rules. My [suggestion](https://github.com/rust-lang/rust/pull/138628#pullrequestreview-2750344682) was to limit the `use` keyword (or at least use closures...) to Rust 2021 so as to avoid having to think about how it interacts with earlier capture rules (as well as potential migrations). I believe this follows from the Edition axiom that [Editions are meant to be adopted](https://rust-lang.github.io/rfcs/3085-edition-2021.html#editions-are-meant-to-be-adopted).

There is an interesting tension with [Rust should feel like one language](https://rust-lang.github.io/rfcs/3085-edition-2021.html#rust-should-feel-like-one-language). My feeling is that there is a missing tenet: the reason we do editions and not fine-grained features is because we wish to avoid combianotoric explosion, where odd combinations of features can lead to untested scenarios. But that is exactly what would be happening here if we allow `use` on older editions. So I think the rule should be that you make new features available on older editions *up until the point where they interact with something that changed* -- in this case, `use` closures interact with the closure capture rules which changed in Rust 2021, so we should limit this feature to Rust 2021 and newer.

Put another way, you should never have to go back and modify an edition migration to work differently. That suggestions you are attempting to push the feature too far back.

**Comment by spastorino** [comment]
We've modified codegen so that we guarantee that `x.use` will do a copy if `X: Copy` is true after monomorphization. Before this change the desugaring to `clone` occured only before monomorphization and hence it would call the `clone` method even for those instances where `X` is a `Copy` type. So with this modification we avoid such situation.

We are not working on convert `x.use` to a move rather than a clone if this is a last-use.

**Comment by spastorino** [comment]
We're currently working on the last-use optimization. We've the liveness analysis needed implemented and we need to extensively test it.

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by nikomatsakis** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2025h2

**Comment by tomassedovic** [comment]
This is a continuing project goal, and the updates below this comment will be for the new period 2026

**Comment by nikomatsakis** [comment]
### Move expressions

RFC https://github.com/rust-lang/rfcs/pull/3968 has been opened.

@TaKO8Ki has done a prototype implementation (tracking issue https://github.com/rust-lang/rust/issues/155050) and it now supports closures but not yet async futures.

### Share trait

P_L1 opened [rust-lang/rust#156828](https://github.com/rust-lang/rust/issues/156828) and [rust-lang/rust#157655](https://github.com/rust-lang/rust/issues/157655) that:

* Added a `Share` trait living besides Clone, uses the separate `share_trait` gate, stays out of the prelude, and has initial impls for `Rc`, `Arc`, `Sender`, and `SyncSender`.
* Adds a `final fn share` that is equivalent to `Clone` and captures the semantic distinction of Share means clone-as-alias, not merely cheap clone.

I am what the best procedural next step is here -- should we have an RFC for `Share`? It seems useful to have in order to document the rationale and why it's useful.
